hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestReplicationManagerScenarios.java

```diff
@@ -0,0 +1,783 @@
+    containerReplicaPendingOps = new ContainerReplicaPendingOps(clock);
+
+    Mockito.when(containerManager.getContainerReplicas(Mockito.any(ContainerID.class))).thenAnswer(
+        invocation -> {
+          ContainerID cid = invocation.getArgument(0);
+          return containerReplicaMap.get(cid);
+        });
+
+    Mockito.when(containerManager.getContainers()).thenAnswer(
+        invocation -> new ArrayList<>(containerInfoSet));
+
+    Mockito.when(nodeManager.getNodeStatus(any(DatanodeDetails.class)))
+        .thenAnswer(invocation -> {
+          DatanodeDetails dn = invocation.getArgument(0);
+          return NODE_STATUS_MAP.getOrDefault(dn, NodeStatus.inServiceHealthy());
+        });
+
+    final HashMap<SCMCommandProto.Type, Integer> countMap = new HashMap<>();
+    for (SCMCommandProto.Type type : SCMCommandProto.Type.values()) {
+      countMap.put(type, 0);
+    }
+    Mockito.when(
+        nodeManager.getTotalDatanodeCommandCounts(any(DatanodeDetails.class),
+            any(SCMCommandProto.Type.class), any(SCMCommandProto.Type.class)))
+        .thenReturn(countMap);
+
+    // Ensure that RM will run when asked.
+    Mockito.when(scmContext.isLeaderReady()).thenReturn(true);
+    Mockito.when(scmContext.isInSafeMode()).thenReturn(false);
+    containerReplicaMap = new HashMap<>();
+    containerInfoSet = new HashSet<>();
+    repReport = new ReplicationManagerReport();
+    repQueue = new ReplicationQueue();
+    ORIGINS.clear();
+    DATANODE_ALIASES.clear();
+    NODE_STATUS_MAP.clear();
+  }
+
+  private ReplicationManager createReplicationManager() throws IOException {
+    return new ReplicationManager(
+        configuration,
+        containerManager,
+        ratisPlacementPolicy,
+        ecPlacementPolicy,
+        eventPublisher,
+        scmContext,
+        nodeManager,
+        clock,
+        legacyReplicationManager,
+        containerReplicaPendingOps) {
+      @Override
+      protected void startSubServices() {
+        // do not start any threads for processing
+      }
+    };
+  }
+
+  protected static UUID getOrCreateOrigin(String origin) {
+    return ORIGINS.computeIfAbsent(origin, (k) -> UUID.randomUUID());
+  }
+
+  private static Stream<Scenario> getTestScenarios() {
+    return TEST_SCENARIOS.stream();
+  }
+
+  private void loadPendingOps(ContainerInfo container, Scenario scenario) {
+    for (PendingReplica r : scenario.getPendingReplicas()) {
+      if (r.getType() == ContainerReplicaOp.PendingOpType.ADD) {
+        containerReplicaPendingOps.scheduleAddReplica(container.containerID(), r.getDatanodeDetails(),
+            r.getReplicaIndex(), Long.MAX_VALUE);
+      } else if (r.getType() == ContainerReplicaOp.PendingOpType.DELETE) {
+        containerReplicaPendingOps.scheduleDeleteReplica(container.containerID(), r.getDatanodeDetails(),
+            r.getReplicaIndex(), Long.MAX_VALUE);
+      }
+    }
+  }
+
+  @ParameterizedTest
+  @MethodSource("getTestScenarios")
```

Comment:
Can we define the test name here too? Right now the default is the full path to the file, which looks pretty verbose.

Reply:
The test name comes from calling `toString()` on the `Scenario` object:

```
public String toString() {
      return resourceName + ": " + description;
    }
```

Originally I just used the test description, but then I worried there can be multiple tests with the same description in different files, and also if there is a failure, its hard to figure out which file the test came from. So I changed it to include the fill path, as it was easier when debugging failures.